### PR TITLE
Enabling constraints when no jacobian is given, tageting issue #19

### LINKED ISSUE #19: FunctionWithApproxJacobian is not defined when calling ipopt_minimize
`ipopt_minimize` will raise an exception when called `NameError: name 'FunctionWithApproxJacobian' is not defined`

It appears there is divergence between the [Optpy](https://github.com/matthias-k/optpy/blob/master/optpy/ipopt_wrapper.py) project and [cyipopt](https://github.com/matthias-k/cyipopt/blob/master/ipopt/ipopt_wrapper.py), with the implementation here ported for Python3 compatibility. However, its missing the related file `jacobian.py` so the wrapper just doesn't work.

I'm trying to (naively) replace a `scipy.optimize.minimize` calls with IPOPT. Figure it should be a working feature since its at the beginning of the README.

The solution doesn't look too bad. May need to merge some changes between Optpy and here, and include a Python2/3 compatible version of `jacobian.py`.

### PR DESCRIPTION
**Pull request by andr1976**
Fixes #19

### CODE REVIEW

**Comment by moorepants** [comment]
@andr1976 Thanks! Could you add a unit test that ensures this functions? You could take this simple example:

```
from scipy.optimize import rosen, rosen_der
from cyipopt import minimize_ipopt
x0 = [1.3, 0.7, 0.8, 1.9, 1.2]
res = minimize_ipopt(rosen, x0, jac=rosen_der)
```

and don't pass the jacobian function to show it works.

File: cyipopt/scipy_interface.py [hunk 1]
```diff
@@ -59,18 +58,20 @@ def __init__(self, fun, args=(), kwargs=None, jac=None, hess=None, hessp=None,
         self._constraint_funs = []
         self._constraint_jacs = []
         self._constraint_args = []
+        self._constraint_kwargs = []
```
Comment (moorepants) [inline cyipopt/scipy_interface.py:61]: What's the purpose of adding this and the other lines with `con_kwargs`? It seems unrelated to the fix.

**Comment by moorepants** [comment]
It would also be helpful if a unit test showed that the args and kwargs are passed through correctly.

**Comment by moorepants** [comment]
Here is the relevant test file: https://github.com/mechmotum/cyipopt/blob/master/cyipopt/tests/unit/test_scipy_optional.py

**Comment by moorepants** [comment]
Does this fix the example here: https://github.com/mechmotum/cyipopt/issues/19#issuecomment-463482188? If so, that could be a test.

**Comment by andr1976** [comment]
Yes fixes example in #19. Would you suggest just expanding the def test_minimize_ipopt_if_scipy(): with more tests, or make a new separate test-function. 
Regarding the args/kwargs passed to the jac lambda wrapping fprime, is ensuring that the objective fun also gets these args/kwargs passed. I did not test that part specifically, but made sure it was consistent and complete. I will make a general test, and see if I can come up with a good test of the args/kwargs

**Comment by brocksam** [comment]
Hi @andr1976, just weighing in here as I wrote most of the CyIpopt tests.

> Would you suggest just expanding the def test_minimize_ipopt_if_scipy(): with more tests, or make a new separate test-function.

It should absolutely be a separate test function. `test_minimize_ipopt_if_scipy` is there to specifically test that the `minimize_ipopt` function works successfully if _Scipy_ is available. You're testing functionality if no Jacobian is supplied so worth having a new test called something like `test_minimize_ipopt_no_jacobian`.

File: cyipopt/tests/unit/test_scipy_optional.py [hunk 2]
```diff
@@ -1,39 +1,68 @@
...
+    assert np.isclose(res.get("fun") , 0)
+    assert res.get("status") == 0
+    assert res.get("success") == True
+    np.testing.assert_allclose(res.get("x"), np.ones(5))
+
+
+@pytest.mark.skipif("scipy" not in sys.modules,
+                    reason="Test only valid if Scipy available.")
+def test_minimize_ipopt_nojac_if_scipy():
+    """If SciPy is installed `minimize_ipopt` works correctly."""
```
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:45]: Can you make the doctoring descriptive of what the test is actually testing?

File: cyipopt/tests/unit/test_scipy_optional.py [hunk 3]
```diff
@@ -1,39 +1,68 @@
...
+    assert isinstance(res, dict)
+    assert np.isclose(res.get("fun") , 0)
+    assert res.get("status") == 0
+    assert res.get("success") == True
+    np.testing.assert_allclose(res.get("x"), np.ones(5),rtol=1e-5)
+
+@pytest.mark.skipif("scipy" not in sys.modules,
+                    reason="Test only valid if Scipy available.")
+def test_minimize_ipopt_nojac_constraints_if_scipy():
+    """If SciPy is installed `minimize_ipopt` works correctly."""
```
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:59]: Same as for other docstring.

File: cyipopt/tests/unit/test_scipy_optional.py [hunk 4]
```diff
@@ -1,39 +1,68 @@
...
+    np.testing.assert_allclose(res.get("x"), np.ones(5))
+
+
+@pytest.mark.skipif("scipy" not in sys.modules,
+                    reason="Test only valid if Scipy available.")
+def test_minimize_ipopt_nojac_if_scipy():
+    """If SciPy is installed `minimize_ipopt` works correctly."""
+    from scipy.optimize import rosen
+    x0 = [1.3, 0.7, 0.8, 1.9, 1.2]
+    options={'tol':1e-7}
```
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:48]: Non-PEP8, style does not match rest of file. Should be

```python
options = {"tol": 1e-7}
```

File: cyipopt/tests/unit/test_scipy_optional.py [hunk 5]
```diff
@@ -1,39 +1,68 @@
...
+    res = cyipopt.minimize_ipopt(rosen, x0,options=options)
```
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:49]: Non-PEP8, should be 

```python
res = cyipopt.minimize_ipopt(rosen, x0, options=options)
```

**Comment by andr1976** [comment]
I have made a few test functions to test that numerical jac via fprime works with and without constraints. Regarding *args and **kwargs I think a general test for objective function *args/**kwargs shall come first, and then the constrint function *args/**kwargs. I can have it on my to-do list, but likely it wont happend in a short while

File: cyipopt/tests/unit/test_scipy_optional.py [hunk 6]
```diff
@@ -1,39 +1,68 @@
...
+    assert isinstance(res, dict)
+    assert np.isclose(res.get("fun") , 0)
+    assert res.get("status") == 0
+    assert res.get("success") == True
```
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:53]: Idiomatically should be checking for identity with booleans not equality here (although I appreciate that that is a slip carried forward from other already written tests!)

```python
assert res.get("success") is True
```

File: cyipopt/tests/unit/test_scipy_optional.py [hunk 7]
```diff
@@ -1,39 +1,68 @@
...
+    np.testing.assert_allclose(res.get("x"), np.ones(5),rtol=1e-5)
+
```
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:55]: Two line spaces between functions.

File: cyipopt/tests/unit/test_scipy_optional.py [hunk 8]
```diff
@@ -1,39 +1,68 @@
...
+    """If SciPy is installed `minimize_ipopt` works correctly."""
+    from scipy.optimize import rosen
+    x0 = [1.3, 0.7, 0.8, 1.9, 1.2]
+    constr={'fun':lambda x: rosen(x)-1.0,'type':'ineq'}
+    res = cyipopt.minimize_ipopt(rosen, x0,constraints=constr)
+    assert isinstance(res, dict)
+    assert np.isclose(res.get("fun") , 1)
+    assert res.get("status") == 0
+    assert res.get("success") == True
+    np.testing.assert_allclose(res.get("x"), np.array([1.001867  , 0.99434067, 1.05070075, 1.17906312, 1.38103001]))
```
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:68]: New line at EoF.

File: cyipopt/tests/unit/test_scipy_optional.py [hunk 9]
```diff
@@ -1,39 +1,68 @@
...
+    assert res.get("success") == True
+    np.testing.assert_allclose(res.get("x"), np.ones(5),rtol=1e-5)
+
+@pytest.mark.skipif("scipy" not in sys.modules,
+                    reason="Test only valid if Scipy available.")
+def test_minimize_ipopt_nojac_constraints_if_scipy():
+    """If SciPy is installed `minimize_ipopt` works correctly."""
+    from scipy.optimize import rosen
+    x0 = [1.3, 0.7, 0.8, 1.9, 1.2]
+    constr={'fun':lambda x: rosen(x)-1.0,'type':'ineq'}
```
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:62]: Non-PEP8

```python
constr = {"fun": lambda x: rosen(x) - 1.0, "type": "ineq"}
```

File: cyipopt/tests/unit/test_scipy_optional.py [hunk 10]
```diff
@@ -1,39 +1,68 @@
...
+    res = cyipopt.minimize_ipopt(rosen, x0,constraints=constr)
```
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:63]: Formatting spaces around comma

```python
res = cyipopt.minimize_ipopt(rosen, x0, constraints=constr)
```

File: cyipopt/tests/unit/test_scipy_optional.py [hunk 11]
```diff
@@ -1,39 +1,68 @@
...
+    assert isinstance(res, dict)
+    assert np.isclose(res.get("fun") , 1)
```
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:65]: Formatting spaces around comma

```python
assert np.isclose(res.get("fun"), 1)
```

File: cyipopt/tests/unit/test_scipy_optional.py (hunk 8 above)
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:68]: This is a long line, can you assign the expected result to a variable to reduce?

```python
expected_res = np.array([1.001867, 0.99434067, 1.05070075, 1.17906312, 1.38103001])
np.testing.assert_allclose(res.get("x"), expected_res)
```

**Review by brocksam — CHANGES_REQUESTED** [review]
If the tests pass then that looks great. Some reformatting to keep the new stuff in line with the style of the rest of the codebase would be great! :)

File: cyipopt/tests/unit/test_scipy_optional.py (hunk 5 above)
Comment (moorepants) [reply cyipopt/tests/unit/test_scipy_optional.py:49]: For these simple changes, you can make "suggestions" with the comment box and we can just merge them in.

File: cyipopt/tests/unit/test_scipy_optional.py (hunk 4 above)
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:48]:
```suggestion
    options = {"tol": 1e-7}
```

File: cyipopt/tests/unit/test_scipy_optional.py (hunk 5 above)
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:49]:
```suggestion
    res = cyipopt.minimize_ipopt(rosen, x0, options=options)

```

File: cyipopt/tests/unit/test_scipy_optional.py (hunk 6 above)
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:53]:
```suggestion
    assert res.get("success") is True
```

File: cyipopt/tests/unit/test_scipy_optional.py (hunk 7 above)
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:55]:
```suggestion


```

File: cyipopt/tests/unit/test_scipy_optional.py (hunk 9 above)
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:62]:
```suggestion
    constr = {"fun": lambda x: rosen(x) - 1.0, "type": "ineq"}
```

File: cyipopt/tests/unit/test_scipy_optional.py (hunk 10 above)
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:63]:
```suggestion
    res = cyipopt.minimize_ipopt(rosen, x0, constraints=constr)
```

File: cyipopt/tests/unit/test_scipy_optional.py (hunk 11 above)
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:65]:
```suggestion
    assert np.isclose(res.get("fun"), 1)
```

File: cyipopt/tests/unit/test_scipy_optional.py (hunk 8 above)
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py]:
```suggestion
    expected_res = np.array([1.001867, 0.99434067, 1.05070075, 1.17906312, 1.38103001])
    np.testing.assert_allclose(res.get("x"), expected_res)

```

File: cyipopt/tests/unit/test_scipy_optional.py [hunk 12]
```diff
@@ -1,39 +1,68 @@
...
+                    reason="Test only valid if Scipy available.")
+def test_minimize_ipopt_if_scipy():
+    """If SciPy is installed `minimize_ipopt` works correctly."""
+    from scipy.optimize import rosen, rosen_der
+    x0 = [1.3, 0.7, 0.8, 1.9, 1.2]
+    res = cyipopt.minimize_ipopt(rosen, x0, jac=rosen_der)
+    assert isinstance(res, dict)
+    assert np.isclose(res.get("fun") , 0)
+    assert res.get("status") == 0
+    assert res.get("success") == True
```
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:38]:
```suggestion
    assert res.get("success") is True
```

File: cyipopt/tests/unit/test_scipy_optional.py (hunk 2 above)
Comment (brocksam) [inline cyipopt/tests/unit/test_scipy_optional.py:45]: E.g.
```suggestion
    """`minimize_ipopt` works correctly on unconstrained problem
    with no Jacobian supplied."""
```

**Review by brocksam — CHANGES_REQUESTED** [review]

**Comment by andr1976** [comment]
Alright, things went a little fast and didn't notice all you suggestions, but I think I have captured most of it, maybe the docstrings for the test functions needs a round more

**Comment by andr1976** [comment]
My editor left some decoration code from the git integration in the text file, hopefully it builds now

File: cyipopt/scipy_interface.py [hunk 13]
```diff
@@ -1,239 +1,240 @@
...
-        x = x[0]
-
-    return OptimizeResult(x=x, success=info['status'] == 0, status=info['status'],
-                          message=info['status_msg'],
-                          fun=info['obj_val'],
-                          info=info,
-                          nfev=problem.nfev,
-                          njev=problem.njev,
-                          nit=problem.nit)
+#!/usr/bin/env python
```
Comment (moorepants) [inline cyipopt/scipy_interface.py:1]: Do you have different line endings on this file? Git seems to be registering that all the lines are changed, instead of just the new ones.

Comment (andr1976) [reply cyipopt/scipy_interface.py:1]: Strange, using VScode, I will check

Comment (andr1976) [reply cyipopt/scipy_interface.py:1]: Arggh, set to CRLF, not LF, I am sure I have set this right earlier - then again I use Visual Studio for the only other project where I commit code to Git/GitHub, so I might have missed it. The diffs in VSCODE looked right, only highlighting lines with actual changes. What to do?

Comment (andr1976) [reply cyipopt/scipy_interface.py:1]: See, but not cheking again I can see whole blocks with no changes also hightlighted as changes
image

Comment (moorepants) [reply cyipopt/scipy_interface.py:1]: You can set the git setting to convert line endings on commit: https://docs.github.com/en/free-pro-team@latest/github/using-git/configuring-git-to-handle-line-endings

Comment (andr1976) [reply cyipopt/scipy_interface.py:1]: What about the fork/branch already in pull request?

Comment (moorepants) [reply cyipopt/scipy_interface.py:1]: I think you'd make the setting and push a new commit. I should fix the files. But maybe you'd have to edit the file.

Comment (andr1976) [reply cyipopt/scipy_interface.py]: I think we are getting there, awaiting the buil check

Comment (andr1976) [reply cyipopt/scipy_interface.py]: Let me know  if more changes are required - sadly IPOPT wasn't ideal for my problem this time (black-box with numerical derivatives), but at least now I have it in the python toolbox, thanks a lot for making this interface.

**Comment by moorepants** [comment]
I added a bit of formatting adjustments and a new docstring as I was learning what everything did. Once tests pass, I'll merge. Thanks @andr1976!

**Comment by moorepants** [comment]
Ok, it's in!
